src/main/java/com/amazon/dlic/auth/ldap/util/LdapHelper.java

```diff
@@ -34,37 +34,46 @@
 import org.ldaptive.referral.SearchReferralHandler;
 
 import org.opensearch.SpecialPermission;
+import org.opensearch.common.settings.Settings;
 
 public class LdapHelper {
 
     private static SearchFilter ALL = new SearchFilter("(objectClass=*)");
+
     @SuppressWarnings("removal")
     public static List<LdapEntry> search(final Connection conn, final String unescapedDn, SearchFilter filter,
-            final SearchScope searchScope, final String[] returnAttributes) throws LdapException {
+            final SearchScope searchScope, final String[] returnAttributes, Settings settings) throws LdapException {
 
         final SecurityManager sm = System.getSecurityManager();
 
         if (sm != null) {
             sm.checkPermission(new SpecialPermission());
         }
-
         try {
             final String baseDn = escapeDn(unescapedDn);
             return AccessController.doPrivileged(new PrivilegedExceptionAction<List<LdapEntry>>() {
                 @Override
                 public List<LdapEntry> run() throws Exception {
                     final List<LdapEntry> entries = new ArrayList<>();
                     final SearchRequest request = new SearchRequest(baseDn, filter);
-                    request.setReferralHandler(new SearchReferralHandler());
+
                     request.setSearchScope(searchScope);
                     request.setDerefAliases(DerefAliases.ALWAYS);
                     request.setReturnAttributes(returnAttributes);
                     final SearchOperation search = new SearchOperation(conn);
-                    // referrals will be followed to build the response
-                    final Response<SearchResult> r = search.execute(request);
-                    final org.ldaptive.SearchResult result = r.getResult();
-                    entries.addAll(result.getEntries());
-                    return entries;
+
+                    if (settings.getAsBoolean(ConfigConstants.FOLLOW_REFERRALS, ConfigConstants.FOLLOW_REFERRALS_DEFAULT)) {
+                        request.setReferralHandler(new SearchReferralHandler());
+                        final Response<SearchResult> r = search.execute(request);
+                        // referrals will be followed to build the response
+                        final org.ldaptive.SearchResult result = r.getResult();
+                        entries.addAll(result.getEntries());
+                        return entries;
+                    } 
+                    else{
+
+                        return entries;
```

Comment:
This doesn't look correct if `r.getResults()` wasn't called

Reply:
This functions to return an empty response when `shouldFollowReferrals`  is `false` and the search would require referral following. I am not sure what the best behavior would be as the use cases leading to this request are so diverse. 

If you had suggestions on what you think should be returned in this case they would be appreciated.

Reply:
Modified so that it now returns a singleton list with an Ldap entry that is `dn="cn=couldNotFollowReferral"` this can be modified by users as they see fit but allows for testing and provides a concise message about the search query.

Reply:
I think the code should look like this with the referral handler unset unless `shouldFollowReferrals` is enabled.  What do you think of this?
```java
                        if (shouldFollowReferrals) {
                           request.setReferralHandler(new SearchReferralHandler());  
                        }

                        final Response<SearchResult> r = search.execute(request);
                        // referrals will be followed to build the response
                        final org.ldaptive.SearchResult result = r.getResult();
                        entries.addAll(result.getEntries());
                        return entries;
```